File: src/main/java/net/theevilreaper/vulpes/generator/generation/java/FontGenerator.java
package net.theevilreaper.vulpes.generator.generation.java;

import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.CodeBlock;
import com.squareup.javapoet.FieldSpec;
import jakarta.inject.Inject;
import jakarta.inject.Singleton;
import net.theevilreaper.vulpes.api.model.FontModel;
import net.theevilreaper.vulpes.api.repository.FontRepository;
import net.theevilreaper.vulpes.font.FontSymbol;
import net.theevilreaper.vulpes.generator.generation.AbstractCodeGenerator;
import net.theevilreaper.vulpes.generator.generation.JavaStructure;
import org.jetbrains.annotations.NotNull;

import java.nio.file.Path;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Singleton
public class FontGenerator extends AbstractCodeGenerator<FontModel> implements JavaStructure {

    private final FontRepository fontRepository;

    @Inject
    public FontGenerator(FontRepository fontRepository) {
        super("FontRegistry", BASE_PACKAGE + ".font");
        this.fontRepository = fontRepository;
    }

    @Override
    protected @NotNull List<FontModel> getModels() {
        return this.fontRepository.findAll();
    }

    @Override
    public void generate(@NotNull Path javaPath) {
        var models = getModels();

        if (models.isEmpty()) return;

        var fieldSpecs = generateFonts(models).values();
        addClassModifiers(this.classBuilder);
        addJetbrainsAnnotation(this.classBuilder);
        addPrivateDefaultConstructor(this.classBuilder);
        addDefaultSuppressAnnotation(this.classBuilder);
        this.classBuilder.addFields(fieldSpecs);
    }

    private @NotNull Map<String, FieldSpec> generateFonts(@NotNull List<FontModel> models) {
        var fields = new HashMap<String, FieldSpec>();
        ClassName fontClass = ClassName.get(FontSymbol.class);
        for (FontModel model : models) {
            if (model.getName() == null || model.getName().isEmpty() || model.getChars().isEmpty()) continue;

            var fontCode = CodeBlock.builder();

            fontCode.add("\\$T.builder()", FontSymbol.class);
            fontCode.add(".symbols(\\$L)", model.getChars().toArray());

            if (model.getAscent() != 0) {
                fontCode.add(".ascent(\\$L)", model.getAscent());
            }

            if (model.getHeight() != 0) {
                fontCode.add(".height(\\$L)", model.getHeight());
            }

            if (model.getShift() != null && !model.getShift().isEmpty()) {
                fontCode.add(".shift(\\$L)", model.getShift());
            }

            FieldSpec fieldValue = FieldSpec.builder(fontClass, model.getName()).build();
            fields.put(model.getName(), fieldValue);
        }

        return fields;
    }

    @Override
    public @NotNull String getName() {
        return "FontGenerator";
    }
}

File: src/main/kotlin/net.theevilreaper.vulpes.generator/generation/java/FontGenerator.kt
package net.theevilreaper.vulpes.generator.generation.java

import com.squareup.javapoet.CodeBlock
import com.squareup.javapoet.FieldSpec
import net.reaper.vulpes.font.FontSymbol
import net.theevilreaper.vulpes.api.model.FontModel
import net.theevilreaper.vulpes.api.repository.FontRepository
import net.theevilreaper.vulpes.api.util.getCharsAsArray
import net.theevilreaper.vulpes.api.util.getShiftAsArray
import net.theevilreaper.vulpes.api.util.hasFontSymbols
import net.theevilreaper.vulpes.api.util.hasShiftData
import net.theevilreaper.vulpes.generator.generation.BaseGenerator
import net.theevilreaper.vulpes.generator.util.BASE_PACKAGE
import net.theevilreaper.vulpes.generator.util.JavaGenerationHelper
import net.theevilreaper.vulpes.generator.util.toVariableString
import org.springframework.stereotype.Service
import java.nio.file.Path

/**
 * @author theEvilReaper
 * @version 1.0.0
 * @since
 **/
@Service
class FontGenerator(
    val fontRepository: FontRepository
) : JavaGenerationHelper, BaseGenerator<FontModel>(
    className = "FontRegistry",
    packageName = "$BASE_PACKAGE.font",
) {

    override fun generate(outputFolder: Path) {
        val models = getModels()

        if (models.isEmpty()) return

        val fieldSpecs = generateFonts(models).values
        this.classSpec.addJavadoc(defaultDocumentation)
        addClassModifiers(this.classSpec)
        addJetbrainsAnnotation(this.classSpec)
        addPrivateDefaultConstructor(this.classSpec)
        this.classSpec.addFields(fieldSpecs)
    }

    private fun generateFonts(models: List<FontModel>): Map<String, FieldSpec> {
        val generatedFonts: MutableMap<String, FieldSpec> = mutableMapOf()

        models.filter { it.name.orEmpty().isNotEmpty() }.forEach { model ->
            if (generatedFonts.isNotEmpty() && generatedFonts.containsKey(model.name)) return@forEach
            if (!model.hasFontSymbols()) return@forEach

            val fontCode = CodeBlock.builder()

            fontCode.add("\$T.builder()", FontSymbol::class.java)
            fontCode.add(".symbols(\$L)", model.getCharsAsArray())

            if (model.ascent != null) {
                fontCode.add(".ascent(\$L)", model.ascent!!)
            }

            if (model.height != null) {
                fontCode.add(".height(\$L)", model.height!!)
            }

            if (model.hasShiftData()) {
                fontCode.add(".shift(\$L)", model.getShiftAsArray())
            }

            val fieldValue = FieldSpec.builder(FontSymbol::class.java, model.name!!.toVariableString()).build()
            generatedFonts[model.name!!] = fieldValue

        }
        return generatedFonts
    }

    override fun getName(): String = "FontGenerator"

    override fun getModels(): List<FontModel> {
        return fontRepository.findAll()
    }
}